src/sync/sync_user.hpp

```diff
@@ -25,12 +25,27 @@
 #include <vector>
 #include <mutex>
 
+#include "../util/atomic_shared_ptr.hpp"
+
 #include <realm/util/optional.hpp>
 
 namespace realm {
 
 class SyncSession;
 
+// A superclass that bindings can inherit from in order to store information
+// upon a `SyncUser` object.
+class SyncUserContext {
+public:
+    virtual ~SyncUserContext() = default;
+};
+
+// A superclass that represents a way for a sync user to build binding contexts.
+class SyncUserContextFactory {
```

Comment:
Any reason to use a base class for this rather than `std::function<std::shared_ptr<SyncUserContext>()>`?